File: src/main/java/io/springreact/jsc/Html.java
package io.springreact.jsc;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The Java "JSX" DSL. Static factories build a {@link UiNode} tree that the universal
 * React client renders. Import statically:
 *
 * <pre>{@code
 * import static io.springreact.jsc.Html.*;
 *
 * div(cls("card"),
 *     h1("Hello " + name),
 *     button(onClick("increment"), "Count: " + count),
 *     ul(items.stream().map(i -> li(i.name())).toList()));
 * }</pre>
 *
 * <p>Each element factory accepts a mix of {@link Attr} (props/events), {@link UiNode}
 * (children), {@link CharSequence} (auto-wrapped as text) and {@link Iterable}/arrays of
 * those — so you can splice mapped lists straight in.
 */
public final class Html {

    private Html() {
    }

    // --- elements ---------------------------------------------------------------

    public static Element el(String tag, Object... parts) {
        Element e = new Element(tag);
        for (Object part : parts) {
            add(e, part);
        }
        return e;
    }

    public static Element div(Object... p)     { return el("div", p); }
    public static Element span(Object... p)    { return el("span", p); }
    public static Element section(Object... p) { return el("section", p); }
    public static Element header(Object... p)  { return el("header", p); }
    public static Element main(Object... p)    { return el("main", p); }
    public static Element h1(Object... p)      { return el("h1", p); }
    public static Element h2(Object... p)      { return el("h2", p); }
    public static Element h3(Object... p)      { return el("h3", p); }
    public static Element p(Object... p)       { return el("p", p); }
    public static Element strong(Object... p)  { return el("strong", p); }
    public static Element em(Object... p)      { return el("em", p); }
    public static Element code(Object... p)    { return el("code", p); }
    public static Element a(Object... p)       { return el("a", p); }
    public static Element button(Object... p)  { return el("button", p); }
    public static Element ul(Object... p)      { return el("ul", p); }
    public static Element li(Object... p)      { return el("li", p); }
    public static Element label(Object... p)   { return el("label", p); }
    public static Element input(Object... p)   { return el("input", p); }
    public static Element form(Object... p)    { return el("form", p); }
    public static Element br()                 { return el("br"); }

    public static Text text(Object value) {
        return new Text(String.valueOf(value));
    }

    /**
     * Escape hatch: render a custom client React component (registered via
     * {@code registerWidget("Name", Comp)}) inside a Java-authored screen. Logic stays on
     * the server; the widget gets the given props plus a {@code call} to invoke actions.
     *
     * <pre>{@code widget("StarRating", attr("value", rating), attr("action", "rate"))}</pre>
     */
    public static Element widget(String name, Object... parts) {
        Element e = new Element("$widget");
        e.props.put("name", name);
        for (Object part : parts) {
            add(e, part);
        }
        return e;
    }

    // --- attributes -------------------------------------------------------------

    public static Attr cls(String value)             { return e -> e.props.put("className", value); }
    public static Attr attr(String key, Object value) { return e -> e.props.put(key, value); }
    public static Attr id(String value)              { return attr("id", value); }
    public static Attr href(String value)            { return attr("href", value); }
    public static Attr type(String value)            { return attr("type", value); }
    public static Attr placeholder(String value)     { return attr("placeholder", value); }
    public static Attr value(Object value)           { return attr("value", value); }
    public static Attr checked(boolean value)        { return attr("checked", value); }
    public static Attr disabled(boolean value)       { return attr("disabled", value); }

    // --- event bindings (resolved to server actions by the client) --------------

    /** Click → call the named server action with the given fixed args. */
    public static Attr onClick(String action, Object... args) {
        return event("onClick", action, null, args);
    }

    /** Change → call the named server action with the given fixed args. */
    public static Attr onChange(String action, Object... args) {
        return event("onChange", action, null, args);
    }

    /** Change → call the action with the input's current value as the first arg. */
    public static Attr onChangeValue(String action, Object... args) {
        return event("onChange", action, "value", args);
    }

    private static Attr event(String key, String action, String eventField, Object... args) {
        return e -> {
            Map<String, Object> binding = new LinkedHashMap<>();
            binding.put("$action", action);
            if (eventField != null) {
                binding.put("event", eventField);
            }
            if (args != null && args.length > 0) {
                binding.put("args", List.of(args));
            }
            e.props.put(key, binding);
        };
    }

    // --- internals --------------------------------------------------------------

    private static void add(Element e, Object part) {
        if (part == null) {
            return;
        }
        if (part instanceof Attr a) {
            a.apply(e);
        } else if (part instanceof UiNode n) {
            e.children.add(n);
        } else if (part instanceof CharSequence s) {
            e.children.add(new Text(s.toString()));
        } else if (part instanceof Iterable<?> it) {
            for (Object x : it) {
                add(e, x);
            }
        } else if (part instanceof Object[] arr) {
            for (Object x : arr) {
                add(e, x);
            }
        } else {
            e.children.add(new Text(String.valueOf(part)));
        }
    }
}

File: src/main/kotlin/io/springreact/jsc/Html.kt
package io.springreact.jsc

/**
 * The Kotlin "JSX" DSL. Functions build a [UiNode] tree the universal React client renders.
 * Import the members:
 *
 * ```
 * import io.springreact.jsc.Html.*
 *
 * div(cls("card"),
 *     h1("Hello $name"),
 *     button(onClick("increment"), "Count: $count"),
 *     ul(items.map { li(it.name) }))
 * ```
 *
 * Each element factory accepts a mix of [Attr] (props/events), [UiNode] children,
 * [CharSequence] (auto-wrapped as text), and [Iterable]/arrays of those.
 *
 * Declared as an `object` with `@JvmStatic` so it reads naturally from both Kotlin and Java.
 */
object Html {

    // --- elements ---------------------------------------------------------------

    @JvmStatic fun el(tag: String, vararg parts: Any?): Element {
        val e = Element(tag)
        for (part in parts) add(e, part)
        return e
    }

    @JvmStatic fun div(vararg p: Any?) = el("div", *p)
    @JvmStatic fun span(vararg p: Any?) = el("span", *p)
    @JvmStatic fun section(vararg p: Any?) = el("section", *p)
    @JvmStatic fun header(vararg p: Any?) = el("header", *p)
    @JvmStatic fun footer(vararg p: Any?) = el("footer", *p)
    @JvmStatic fun nav(vararg p: Any?) = el("nav", *p)
    @JvmStatic fun main(vararg p: Any?) = el("main", *p)
    @JvmStatic fun h1(vararg p: Any?) = el("h1", *p)
    @JvmStatic fun h2(vararg p: Any?) = el("h2", *p)
    @JvmStatic fun h3(vararg p: Any?) = el("h3", *p)
    @JvmStatic fun p(vararg p: Any?) = el("p", *p)
    @JvmStatic fun strong(vararg p: Any?) = el("strong", *p)
    @JvmStatic fun em(vararg p: Any?) = el("em", *p)
    @JvmStatic fun code(vararg p: Any?) = el("code", *p)
    @JvmStatic fun a(vararg p: Any?) = el("a", *p)
    @JvmStatic fun button(vararg p: Any?) = el("button", *p)
    @JvmStatic fun ul(vararg p: Any?) = el("ul", *p)
    @JvmStatic fun li(vararg p: Any?) = el("li", *p)
    @JvmStatic fun label(vararg p: Any?) = el("label", *p)
    @JvmStatic fun input(vararg p: Any?) = el("input", *p)
    @JvmStatic fun form(vararg p: Any?) = el("form", *p)
    @JvmStatic fun small(vararg p: Any?) = el("small", *p)
    @JvmStatic fun br() = el("br")

    @JvmStatic fun text(value: Any?) = Text(value.toString())

    /** A placeholder a layout renders the current child screen into. */
    @JvmStatic fun slot() = Element("\$slot")

    /**
     * Escape hatch: render a custom client React component (registered via
     * `registerWidget("Name", Comp)`) inside a server-authored screen.
     */
    @JvmStatic fun widget(name: String, vararg parts: Any?): Element {
        val e = Element("\$widget")
        e.props["name"] = name
        for (part in parts) add(e, part)
        return e
    }

    // --- attributes -------------------------------------------------------------

    @JvmStatic fun cls(value: String) = Attr { it.props["className"] = value }
    @JvmStatic fun attr(key: String, value: Any?) = Attr { it.props[key] = value }
    @JvmStatic fun name(value: String) = attr("name", value)
    @JvmStatic fun id(value: String) = attr("id", value)
    @JvmStatic fun href(value: String) = attr("href", value)
    @JvmStatic fun type(value: String) = attr("type", value)
    @JvmStatic fun placeholder(value: String) = attr("placeholder", value)
    @JvmStatic fun value(value: Any?) = attr("value", value)
    @JvmStatic fun checked(value: Boolean) = attr("checked", value)
    @JvmStatic fun disabled(value: Boolean) = attr("disabled", value)

    // --- event bindings (resolved to server actions by the client) --------------

    @JvmStatic fun onClick(action: String, vararg args: Any?) = event("onClick", action, null, false, args)
    @JvmStatic fun onChange(action: String, vararg args: Any?) = event("onChange", action, null, false, args)
    @JvmStatic fun onChangeValue(action: String, vararg args: Any?) = event("onChange", action, "value", false, args)

    /** Submit: the client gathers the form's named fields into an object passed as arg 0. */
    @JvmStatic fun onSubmit(action: String, vararg args: Any?) = event("onSubmit", action, null, true, args)

    private fun event(key: String, action: String, eventField: String?, form: Boolean, args: Array<out Any?>): Attr =
        Attr { e ->
            val binding = LinkedHashMap<String, Any?>()
            binding["\$action"] = action
            if (eventField != null) binding["event"] = eventField
            if (form) binding["form"] = true
            if (args.isNotEmpty()) binding["args"] = args.toList()
            e.props[key] = binding
        }

    // --- internals --------------------------------------------------------------

    private fun add(e: Element, part: Any?) {
        when (part) {
            null -> {}
            is Attr -> part.apply(e)
            is UiNode -> e.children.add(part)
            is CharSequence -> e.children.add(Text(part.toString()))
            is Iterable<*> -> part.forEach { add(e, it) }
            is Array<*> -> part.forEach { add(e, it) }
            else -> e.children.add(Text(part.toString()))
        }
    }
}
